### ISSUE TITLE
Feedback on PyPy specific GC hooks

### ISSUE THREAD

**Issue by thedrow** [cython/cython#2648]
After I read this [blog post](https://morepypy.blogspot.com/2018/09/inside-cpyext-why-emulating-cpython-c.html) about why emulating the CPython API is so hard for PyPy I raised the fact that we could introduce PyPy specific GC hooks for the cpyext layer in PyPy's IRC channel.
These would be used by Cython instead of `Py_INCREF` and `Py_DECREF` etc. to track the lifetime objects.
We're wondering what are the API requirements from Cython's side to implement such a thing.
How do you imagine such API to look like?
How hard would it be to generate code that uses those hooks?
What do you think that should happen if a Cython uses `Py_INCREF` or other memory management macro manually? Should this feature be hidden under a feature flag?

**Comment by scoder** [comment]
It's unclear to me what you mean by "GC hooks". If it's just about replacing incref/decref with something else, then that sounds like something that PyPy could do in its header files. Or is the idea to (somehow) change the lifecycle of objects?

**Comment by thedrow** [comment]
The idea is to provide a different API to interact with the GC so that PyPy would be able to completely avoid reference counting and instead use it's own optimized GC.
So instead of `Py_INCREF` and `Py_DECREF` you'd have another API.
This issue is about collaborating with the Cython community on such API.

**Comment by scoder** [comment]
Ok. So, one of the problems is that Cython almost never has a complete view of the lifetime of an object. It knows when it owns a reference to it, but it often does not know who else owns it. Reference counting is a very convenient way to deal with this situation.

I'd be happy to see proposals how to improve this, but I don't have any myself.

**Comment by thedrow** [comment]
Obviously we need a macro or method to mark the object as tracked by PyPy's GC.
So `PyPy_TRACK` or something similar would interact with the GC.
I don't think you'll need full view of the lifetime of the object.
I think you'll need to notify PyPy if the object ever leaves the current scope, thus moving itself from the nursery to an older generation. I'm no expert though.

**Comment by scoder** [comment]
Could you at least outline your line of thinking or reasoning? It's still very unclear what kind of interaction you're envisioning here. For example, why should Cython use a hypothetical `PyPy_TRACK()` for objects that originate from PyPy in the first place? What GC based interaction could there be that does not involve reference counting? (And, what's wrong with reference counting when w.r.t. PyPy? They've implemented it, right?)

**Comment by thedrow** [comment]
Let's begin with what's wrong with reference counting.
There are two reasons for reference counting slowing down C extensions:

1. PyPy's GC is one of the main reasons PyPy is much faster than CPython.
1. Crossing the border between RPython and C requires translating the object from PyObject* to W_Root and back.

You can read more about it in the blog post I linked to.
TL;DR:

> So, we can see that calling C from RPython introduce some overhead. Can we measure it?
Assuming that the conversion between W_Root and PyObject* has a reasonable cost (as explained by the previous section), the overhead introduced by a single border-cross is still acceptable, especially if the callee is doing some non-negligible amount of work.
However this is not always the case. There are basically three problems that make (or used to make) cpyext super slow:
> 1) Paying the border-crossing cost for trivial operations which are called very often, such as Py_INCREF.
> 2) Crossing the border back and forth many times, even if it's not strictly needed.
> 3) Paying an excessive cost for argument and return value conversions.

Calling Py_INCREF crosses the RPython-C border which causes the slowdowns described in that blog post.
Reference counting is indeed implemented but it is currently emulated and we translate between PyPy's representation of an object and CPython's representation.
If we could avoid that, all C extensions get a major win.

Cython in my opinion should only call `PyPy_TRACK()` whenever it allocates an object in C. All other objects are considered tracked since they were already created in RPython.

The GC needs to be notified when objects allocated in C go out of scope since that's how it decides if they should be moved to an older generation than the nursery and survive the current scope so Cython needs to inject that code as well.

There is a problem with this approach. Using the C API directly becomes a problem so it should either result in a compilation error or in a fallback to reference counting. Another approach is handling it on the RPython side of the border by adding a flag to indicate that this object is not reference counted and thus calling `Py_INCREF` is a no-op in those C API methods.

I won't deny that this is a complicated matter, but I think the entire Python community can benefit from it.
